File: infra/mpv.go
package infra

import (
	"bufio"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"os/exec"
	"sync"
	"time"
)

// MpvClient controls a running mpv instance via its JSON IPC socket.
// All methods are safe for concurrent use.
type MpvClient struct {
	SocketPath string // e.g. /tmp/YTcliV2-mpv.sock
	Binary     string // default: "mpv"

	mu      sync.Mutex
	conn    net.Conn
	reader  *bufio.Scanner
	cmd     *exec.Cmd
	running bool
}

// NewMpvClient creates a client for an mpv instance.
func NewMpvClient(socketPath, binary string) *MpvClient {
	return &MpvClient{
		SocketPath: socketPath,
		Binary:     binary,
	}
}

// mpvRequest is a JSON-RPC command.
type mpvRequest struct {
	Command   []interface{} `json:"command"`
	RequestID int           `json:"request_id,omitempty"`
}

// mpvResponse is a JSON-RPC response.
type mpvResponse struct {
	Error     string      `json:"error"`
	Data      interface{} `json:"data"`
	RequestID int         `json:"request_id"`
}

// Start spawns mpv with --input-ipc-server and waits for the socket.
func (m *MpvClient) Start() error {
	os.Remove(m.SocketPath)

	m.cmd = exec.Command(m.Binary,
		"--idle",
		"--no-terminal",
		"--no-video",
		"--no-audio-display",
		"--volume=100",
		"--volume-max=100",
		"--replaygain=no",
		"--af=scaletempo",
		fmt.Sprintf("--input-ipc-server=%s", m.SocketPath),
	)
	if err := m.cmd.Start(); err != nil {
		return fmt.Errorf("mpv exec: %w", err)
	}
	m.running = true

	// Wait for socket to appear
	for i := 0; i < 50; i++ {
		if _, err := os.Stat(m.SocketPath); err == nil {
			return m.connect()
		}
		time.Sleep(100 * time.Millisecond)
	}
	return fmt.Errorf("mpv socket not ready after 5s")
}

// connect opens the Unix socket connection.
func (m *MpvClient) connect() error {
	conn, err := net.Dial("unix", m.SocketPath)
	if err != nil {
		return fmt.Errorf("mpv dial: %w", err)
	}
	m.conn = conn
	m.reader = bufio.NewScanner(conn)
	return nil
}

// sendCommand sends a JSON command and reads one response line.
// Returns an error if mpv is not connected (nil-conn safe).
func (m *MpvClient) sendCommand(cmd []interface{}) (*mpvResponse, error) {
	m.mu.Lock()
	defer m.mu.Unlock()

	if m.conn == nil {
		return nil, fmt.Errorf("mpv not connected")
	}

	req := mpvRequest{Command: cmd, RequestID: 1}
	data, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("mpv marshal: %w", err)
	}
	if _, err := m.conn.Write(append(data, '\n')); err != nil {
		return nil, fmt.Errorf("mpv write: %w", err)
	}
	if m.reader.Scan() {
		var resp mpvResponse
		if err := json.Unmarshal([]byte(m.reader.Text()), &resp); err != nil {
			return nil, fmt.Errorf("mpv unmarshal: %w", err)
		}
		if resp.Error != "success" && resp.Error != "" {
			return &resp, fmt.Errorf("mpv error: %s", resp.Error)
		}
		return &resp, nil
	}
	return nil, fmt.Errorf("mpv no response")
}

// LoadURL tells mpv to load a stream URL.
func (m *MpvClient) LoadURL(url string) error {
	_, err := m.sendCommand([]interface{}{"loadfile", url, "replace"})
	return err
}

// SetPause pauses or resumes.
func (m *MpvClient) SetPause(paused bool) error {
	_, err := m.sendCommand([]interface{}{"set_property", "pause", paused})
	return err
}

// SetVolume sets volume 0-100.
func (m *MpvClient) SetVolume(level int) error {
	_, err := m.sendCommand([]interface{}{"set_property", "volume", level})
	return err
}

// GetVolume returns the current mpv volume (0-100).
func (m *MpvClient) GetVolume() (int, error) {
	resp, err := m.sendCommand([]interface{}{"get_property", "volume"})
	if err != nil {
		return 0, err
	}
	switch v := resp.Data.(type) {
	case float64:
		return int(v), nil
	}
	return 0, nil
}

// GetPosition returns current position in seconds.
func (m *MpvClient) GetPosition() (float64, error) {
	resp, err := m.sendCommand([]interface{}{"get_property", "time-pos"})
	if err != nil {
		return 0, err
	}
	switch v := resp.Data.(type) {
	case float64:
		return v, nil
	}
	return 0, nil
}

// Seek jumps forward (+seconds) or backward (-seconds).
func (m *MpvClient) Seek(seconds float64) error {
	_, err := m.sendCommand([]interface{}{"seek", seconds, "relative"})
	return err
}

// Stop stops current playback.
func (m *MpvClient) Stop() error {
	_, err := m.sendCommand([]interface{}{"stop"})
	return err
}

// GetDuration returns total duration in seconds.
func (m *MpvClient) GetDuration() (float64, error) {
	resp, err := m.sendCommand([]interface{}{"get_property", "duration"})
	if err != nil {
		return 0, err
	}
	switch v := resp.Data.(type) {
	case float64:
		return v, nil
	}
	return 0, nil
}

// Quit shuts down mpv.
func (m *MpvClient) Quit() error {
	m.sendCommand([]interface{}{"quit"})
	if m.conn != nil {
		m.conn.Close()
	}
	if m.cmd != nil && m.cmd.Process != nil {
		m.cmd.Process.Kill()
		m.cmd.Wait()
	}
	os.Remove(m.SocketPath)
	m.running = false
	return nil
}

File: src/mpv.rs
/// mpv control via Unix socket JSON-RPC.

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::io::{BufRead, BufReader, Write};
use std::os::unix::net::UnixStream;
use std::process::{Child, Command};
use std::sync::Mutex;
use std::time::Duration;

#[derive(Serialize)]
struct Request {
    command: Vec<serde_json::Value>,
    request_id: u32,
}

#[derive(Deserialize)]
struct Response {
    error: Option<String>,
    data: Option<serde_json::Value>,
}

pub struct Mpv {
    socket_path: String,
    binary: String,
    process: Option<Child>,
    conn: Mutex<Option<UnixStream>>,
}

impl Mpv {
    pub fn new(socket_path: &str, binary: &str) -> Self {
        Self {
            socket_path: socket_path.into(),
            binary: binary.into(),
            process: None,
            conn: Mutex::new(None),
        }
    }

    pub fn start(&mut self) -> Result<()> {
        let _ = std::fs::remove_file(&self.socket_path);

        let mut cmd = Command::new(&self.binary);
        cmd.args([
            "--idle",
            "--no-terminal",
            "--no-video",
            "--no-audio-display",
            "--volume=100",
            "--volume-max=100",
            "--replaygain=no",
            "--af=scaletempo",
            &format!("--input-ipc-server={}", self.socket_path),
        ]);

        let child = cmd.spawn().context("gagal start mpv")?;
        self.process = Some(child);

        for _ in 0..50 {
            if std::path::Path::new(&self.socket_path).exists() {
                return self.connect();
            }
            std::thread::sleep(Duration::from_millis(100));
        }

        anyhow::bail!("mpv socket tidak muncul setelah 5s di {}", self.socket_path);
    }

    fn connect(&self) -> Result<()> {
        let stream = UnixStream::connect(&self.socket_path)
            .context("gagal connect ke mpv socket")?;
        stream.set_read_timeout(Some(Duration::from_secs(2)))?;
        *self.conn.lock().unwrap() = Some(stream);
        Ok(())
    }

    fn send(&self, args: &[serde_json::Value]) -> Result<Option<serde_json::Value>> {
        let req = Request {
            command: args.to_vec(),
            request_id: 1,
        };

        let mut json = serde_json::to_string(&req)?;
        json.push('\n');

        let mut guard = self.conn.lock().unwrap();
        let stream = guard.as_mut().context("mpv belum connect")?;

        stream.write_all(json.as_bytes())?;
        stream.flush()?;

        let mut line = String::new();
        let mut reader = BufReader::new(stream.by_ref());
        reader.read_line(&mut line)?;

        if line.is_empty() {
            anyhow::bail!("koneksi mpv terputus");
        }

        let resp: Response = serde_json::from_str(line.trim())?;

        if let Some(err) = resp.error {
            if err != "success" && !err.is_empty() {
                anyhow::bail!("mpv error: {}", err);
            }
        }

        Ok(resp.data)
    }

    // ── High-level API ──

    pub fn load_url(&self, url: &str) -> Result<()> {
        self.send(&["loadfile".into(), url.into(), "replace".into()])?;
        Ok(())
    }

    pub fn set_pause(&self, paused: bool) -> Result<()> {
        self.send(&["set_property".into(), "pause".into(), paused.into()])?;
        Ok(())
    }

    pub fn toggle_pause(&self) -> Result<bool> {
        let data = self.send(&["get_property".into(), "pause".into()])?;
        let current = data.and_then(|v| v.as_bool()).unwrap_or(false);
        self.set_pause(!current)?;
        Ok(!current)
    }

    pub fn set_volume(&self, vol: u8) -> Result<()> {
        let v: u32 = (vol as u32).min(100);
        self.send(&["set_property".into(), "volume".into(), v.into()])?;
        Ok(())
    }

    pub fn get_volume(&self) -> Result<u8> {
        let data = self.send(&["get_property".into(), "volume".into()])?;
        let vol = data
            .and_then(|v| v.as_f64())
            .map(|v| v as u8)
            .context("mpv: volume property returned null")?;
        Ok(vol)
    }

    pub fn get_position(&self) -> Result<f64> {
        let data = self.send(&["get_property".into(), "time-pos".into()])?;
        Ok(data.and_then(|v| v.as_f64()).unwrap_or(0.0))
    }

    pub fn get_duration(&self) -> Result<f64> {
        let data = self.send(&["get_property".into(), "duration".into()])?;
        Ok(data.and_then(|v| v.as_f64()).unwrap_or(0.0))
    }

    pub fn seek(&self, seconds: f64) -> Result<()> {
        self.send(&["seek".into(), seconds.into(), "relative".into()])?;
        Ok(())
    }

    pub fn stop(&self) -> Result<()> {
        let _ = self.send(&["stop".into()]);
        Ok(())
    }

    pub fn get_paused(&self) -> Result<bool> {
        let data = self.send(&["get_property".into(), "pause".into()])?;
        Ok(data.and_then(|v| v.as_bool()).unwrap_or(false))
    }

    pub fn get_eof_reached(&self) -> Result<bool> {
        let data = self.send(&["get_property".into(), "eof-reached".into()])?;
        Ok(data.and_then(|v| v.as_bool()).unwrap_or(false))
    }

    pub fn quit(&mut self) {
        let _ = self.send(&["quit".into()]);
        if let Some(mut child) = self.process.take() {
            let _ = child.kill();
            let _ = child.wait();
        }
        *self.conn.lock().unwrap() = None;
        let _ = std::fs::remove_file(&self.socket_path);
    }
}

impl Drop for Mpv {
    fn drop(&mut self) {
        self.quit();
    }
}
